onnx/onnx.proto

```diff
@@ -186,6 +192,123 @@ message NodeProto {
+  // introducing a new stable function with the same name.
+  optional OperatorStatus status = 3;
+
+  // The inputs and outputs of the function.
+  repeated string input = 4;
+  repeated string output = 5;
+
+  // The attributes of the function.
+  repeated string attribute= 6;
+
+  // The nodes in the function.
+  repeated NodeProto node = 7;
+  // A human-readable documentation for this function. Markdown is allowed.
+  optional string doc_string = 8;
+}
+
+// Training information
+// TrainingInfoProto contains the minimized loss function (e.g., mean
+// squared error), the used optimization algorithm (a mapping from gradient
+// and the current tensors to their new values; that is, one training iteration),
+// and the binding between the optimized tensors and their gradient tensors'
+// names (we need to give a name (e.g., "gradient_of_X") to the gradient of
+// tensor "X", so that gradient tensors can be referenced in the training 
+// algorithm).
+//
+// This design is mainly designed for popular stochastic gradient-based methods.
+// Conceptually, a TrainingInfoProto defines a "partial computation graph"
+// which will be merged into the inference graph to form a full training
+// computation graph.
+message TrainingInfoProto {
+  // Training-specific initializers, for example, accumulated squared gradient
+  // in ADAGRAD optimizer, momentum, or number of executed training iterations.
+  // The values can only be inputs of "loss" and "optimizer" defined below.
+  repeated TensorProto additional_initializer = 1;
+
+  // Training-specific inputs, for example, labels in classification problems.
+  // The values can only be inputs of "loss" and "optimizer" defined below.
+  // The full input list of the training computation graph is the concatenation
+  // of "ModelProto.graph.input" and
+  // "ModelProto.training_info.additional_input".
+  repeated ValueInfoProto additional_input = 2;
+
+  // Training-specific outputs, for example, new tensor values computed by the
+  // specified training algorithm. The full output list of the training
+  // computation graph is the concatenation of "ModelProto.graph.output" and
+  // "ModelProto.training_info.additional_output".
+  repeated ValueInfoProto additional_output = 3;
+  
+  // Loss node which consumes tensors in "ModelProto.graph", "additional_inputs"
+  // , and "additional_initializer". Loss nodes produces a scalar loss value.
+  // In training phase, the optimization algorithm would be launch to minimizes
+  // loss node's output value. "loss" can be an ONNX primitive operator or a
+  // function defined in "ModelProto.function."
+  //
+  // The field MUST be present.
+  optional NodeProto loss = 4;
+
+  // Optimized tensors and their gradient tensor names. Each pair describes a
+  // tensor's name (key) and its gradient's name (value). The gradient tensors
+  // are only visible to the input list of "optimizer."
+  repeated StringStringEntryProto gradient_binding = 5;
```

Comment:
Please explain how the gradient tensors are associated with `loss`.

Reply:
Do we really need this binding? Based on the document https://github.com/onnx/onnx/files/3208156/ONNX.Training.Discussion.pptx, the optimizer has W and Gradient W as input. The name of "gradient W", which is output from the backward prop, seems could be derived from the backend and runtime implementation. The current Pytorch and TF APIs do not support user-defined names for the gradients for the optimizers as a reference.

Reply:
We need it. Pytorch creates such a binding in another (but equivalent) way. They have a dictionary where key is parameter and value is its gradient tensor.